File: src/main/java/com/vc/medium/L47.scala
package com.vc.medium

object L47 {
  def permuteUnique(arr: Array[Int]): List[List[Int]] = {
    import scala.collection.mutable
    val res = new mutable.ListBuffer[List[Int]]()
    val list = new mutable.ArrayBuffer[Int]()
    scala.util.Sorting.quickSort(arr)
    val n = arr.length
    val visited = new Array[Boolean](n)
    def solve(): Unit = {
      if(list.size == n) res += list.toList
      else {
        (0 until n).foreach(i => {
          if(!(visited(i) || (i > 0 && arr(i) == arr(i - 1) && !visited(i - 1)))) {
            visited(i) = true
            list += arr(i)
            solve()
            list.remove(list.size - 1)
            visited(i) = false
          }
        })
      }
    }
    solve()
    res.toList
  }
}

File: src/main/java/com/vc/medium/L47.java
package com.vc.medium;

class L47 {
    List<List<Integer>> res = new ArrayList<>();
    public List<List<Integer>> permuteUnique(int[] nums) {
        Arrays.sort(nums);
        solve(new ArrayList<Integer>(), nums, new boolean[nums.length]);
        return res;
    }

    private void solve(List<Integer> list, int[] arr, boolean[] visited) {
        if(list.size() == arr.length) {
            res.add(new ArrayList<>(list));
        }

        for(int i = 0; i < arr.length; i++) {
            if(visited[i]) continue;
            if(i > 0 && arr[i] == arr[i - 1] && !visited[i - 1]) continue;

            visited[i] = true;
            list.add(arr[i]);
            solve(list, arr, visited);
            list.remove(list.size() - 1);
            visited[i] = false;
        }
    }
}
